File: app/src/main/java/com/david/redcristianauno/PrincipalActivity.java
package com.david.redcristianauno;

import android.app.AlertDialog;
import android.app.DatePickerDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.FragmentManager;

import android.view.MenuItem;


import android.view.Menu;
import android.widget.DatePicker;
import android.widget.TextView;
import android.widget.Toast;


import com.david.redcristianauno.Firestore.InsertarDatos;
import com.david.redcristianauno.Firestore.LeerDatos;
import com.david.redcristianauno.Historico.HistoricoDiarioFragment;
import com.david.redcristianauno.POJOs.HistoricoSemanal;
import com.david.redcristianauno.POJOs.HistoricoSemanalSubred;
import com.david.redcristianauno.POJOs.RegistroCelula;
import com.david.redcristianauno.POJOs.RegistroSubred;
import com.david.redcristianauno.ui.activities.LoginActivity;
import com.david.redcristianauno.ui.fragments.CreateConfigurationDialogFragment;
import com.google.android.material.navigation.NavigationView;
import com.google.firebase.FirebaseApp;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;


public class PrincipalActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener, DatePickerDialog.OnDateSetListener {

    private TextView tv_nombre_usuario, tv_correo_usuario;
    private FirebaseDatabase firebaseDatabase;
    private DatabaseReference databaseReference;
    private List<String> listaSemanal = new ArrayList<String>();
    static int asistencia = 0;
    static int invitados = 0;
    static int ninos = 0;
    static double ofrenda = 0;
    public static String correo_usuario = "";
    private NavigationView navigationView;
    private Menu menu;
    public static String res = "";
    public int resvar;

    long maxid = 0;
    long maxidSubred = 0;

    private LeerDatos l = new LeerDatos();
    private InsertarDatos inda = new InsertarDatos();



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_principal);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = findViewById(R.id.drawer_layout);

        navigationView = findViewById(R.id.nav_view);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        View nombreview = navigationView.getHeaderView(0);
        View correoview = navigationView.getHeaderView(0);

        tv_nombre_usuario = (TextView) nombreview.findViewById(R.id.tv_nombre_usuario);
        tv_correo_usuario = (TextView) correoview.findViewById(R.id.tv_correo);

        navigationView.setNavigationItemSelectedListener(this);

        FragmentManager fragmentManager = getSupportFragmentManager();
        fragmentManager.beginTransaction().replace(R.id.contenedor,new MenuFragment()).commit();

        inicializarFirebase();

        Calendar c = Calendar.getInstance();
        c.get(Calendar.DAY_OF_MONTH);
        c.get(Calendar.MONTH);
        c.get(Calendar.YEAR);
        String fecha = DateFormat.getDateInstance(DateFormat.FULL).format(c.getTime());

        //checar(fecha);

        inda.leerHistorico(fecha);

        correo_usuario = Preferences.obtenerPreferencesString(this, Preferences.PREFERENCES_USUARIO_LOGIN);

        tv_correo_usuario.setText(correo_usuario);

        l.datosSesion(tv_nombre_usuario,correo_usuario, menu, navigationView, PrincipalActivity.this);

        //ocultardatos();
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } /*else {
            super.onBackPressed();
        }*/
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.principal, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();


        if (id == R.id.salida) {
            new AlertDialog.Builder(PrincipalActivity.this)
                    .setTitle("Cerrar Sesión")
                    .setMessage("¿Estás seguro de cerrar la sesión?")
                    .setPositiveButton("Aceptar", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            Preferences.savePreferenceBoolean(PrincipalActivity.this,false,Preferences.PREFENCE_ESTADO_BUTTON_SESION);
                            Intent i = new Intent(PrincipalActivity.this, LoginActivity.class);
                            startActivity(i);
                            PrincipalActivity.this.finish();
                        }
                    })
                    .setNegativeButton("Cancelar", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.cancel();
                        }
                    }).show();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();
        FragmentManager fragmentManager = getSupportFragmentManager();

        if (id == R.id.nav_home) {
            fragmentManager.beginTransaction().replace(R.id.contenedor,new MenuFragment()).commit();
        } else if (id == R.id.nav_gallery) {
            fragmentManager.beginTransaction().replace(R.id.contenedor,new RegistroSubredFragment()).commit();
            resvar = 1;
        } else if (id == R.id.nav_slideshow) {
            fragmentManager.beginTransaction().replace(R.id.contenedor,new RegistroCelulaFragment()).commit();
            resvar=2;
        } else if (id == R.id.nav_tools) {
            fragmentManager.beginTransaction().replace(R.id.contenedor,new HistoricoFragment()).commit();
            resvar=3;
        }else if(id == R.id.nav_biblia){
            fragmentManager.beginTransaction().replace(R.id.contenedor,new BibliaFragment()).commit();
        } else if (id == R.id.nav_share) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, new PermisosUsuariosFragment()).commit();
        }else if(id == R.id.crear){
            fragmentManager.beginTransaction().replace(R.id.contenedor, new CreateConfigurationDialogFragment()).commit();
        }else if(id == R.id.noticias){
            fragmentManager.beginTransaction().replace(R.id.contenedor, new noticiasFragment()).commit();
        }


        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    public void inicializarFirebase(){
        FirebaseApp.initializeApp(this);
        firebaseDatabase = FirebaseDatabase.getInstance();
        databaseReference = firebaseDatabase.getReference();
    }

    private void checar(final String fecha) {
        databaseReference.child("Historico").orderByChild("fecha").addListenerForSingleValueEvent(new ValueEventListener() {
            boolean bandera = true;
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                for (final DataSnapshot snapshot: dataSnapshot.getChildren()){
                    HistoricoSemanal rc = snapshot.getValue(HistoricoSemanal.class);
                    String fecha_celula = rc.getFecha();

                    System.out.println("Fecha historico:" + fecha_celula);
                    if(fecha.equals(fecha_celula)){
                        bandera=false;
                    }
                    System.out.println("Bandera: " +  bandera);
                }

                if (bandera==true){
                    formatoSemanal();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }
    public void formatoSemanal(){
        Calendar miCalendario = Calendar.getInstance();
        miCalendario.setFirstDayOfWeek(Calendar.TUESDAY);

        boolean bandera = false;

        int diaSemana = miCalendario.get(Calendar.DAY_OF_WEEK);

        if(diaSemana!=miCalendario.getFirstDayOfWeek()){
            bandera=false;
        }else{
            bandera=true;
        }

        if(bandera == true){

            int dia_del_mes = miCalendario.get(Calendar.DAY_OF_MONTH);
            System.out.println("El dia es: " + dia_del_mes);
            int mes = miCalendario.get(Calendar.MONTH) + 1;
            System.out.println("El mes es: " +  mes);
            int año = miCalendario.get(Calendar.YEAR);
            System.out.println("El año es: " +  año);

            int nuevo = dia_del_mes;

            String fechas[] = new String[7];

            for (int i = 0; i<7; i++){
                if(nuevo  == 0){
                    nuevo = 1;
                }
                nuevo-=1;

                switch (mes){
                    case 1:
                        if (nuevo <= 0){
                            mes = mes-1;
                            nuevo = 31-nuevo;
                        }
                        break;
                    case 2:
                        if (nuevo <= 0){
                            mes = mes-1;
                            nuevo = 31-nuevo;
                        }
                        break;
                    case 3:
                        if((año%4 == 0) && ((año % 100 !=0) || (año % 400 == 0))){
                            if(nuevo<=0){
                                mes = mes -1;
                                nuevo = 29-nuevo;
                            }
                        }else {
                            if(nuevo<=0){
                                mes = mes -1;
                                nuevo = 28 - nuevo;
                            }
                        }
                        break;
                    case 4:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 31 - nuevo;
                        }
                        break;
                    case 5:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 30 - nuevo;
                        }
                        break;
                    case 6:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 31 - nuevo;
                        }
                        break;
                    case 7:
                        if (nuevo <= 0){
                            mes = mes-1;
                            nuevo = 30 - nuevo;
                        }
                        break;
                    case 8:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 31 - nuevo;
                        }
                        break;
                    case 9:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 31 - nuevo;
                        }
                        break;
                    case 10:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 30 - nuevo;
                        }
                        break;
                    case 11:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 31 - nuevo;
                        }
                        break;
                    case 12:
                        if (nuevo <= 0){
                            mes = mes -1;
                            nuevo = 30 - nuevo;
                        }
                        break;
                }
                miCalendario.set(Calendar.DAY_OF_MONTH, nuevo);
                miCalendario.set(Calendar.MONTH, mes - 1);
                miCalendario.set(Calendar.YEAR, año);
                final String fecha = DateFormat.getDateInstance(DateFormat.FULL).format(miCalendario.getTime());
                System.out.println(fecha);
                fechas[i] = fecha;
            }
            crearFormato(fechas);
            crearFormatoSubred(fechas);
        }
    }



    private void crearFormatoSubred(final String[] fechas) {
        databaseReference.child("Registro Subred").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                int total_asistencia = 0;
                double total_ofrenda = 0;

                for (final DataSnapshot snapshot: dataSnapshot.getChildren()){
                    RegistroSubred rc = snapshot.getValue(RegistroSubred.class);
                    String fecha_celula = rc.getFecha_subred();

                    for (int i = 0; i < fechas.length; i++) {
                        if (fechas[i].equals(fecha_celula)) {
                            total_asistencia = total_asistencia + rc.getAsistencia_subred();
                            total_ofrenda = total_ofrenda + rc.getOfrenda_subred();
                        }
                    }
                }

                totalSemanalSubred(total_asistencia,total_ofrenda);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void totalSemanalSubred(final int total_asistencia, final double total_ofrenda) {
        databaseReference.child("Historico Subred").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()){
                    maxidSubred = (dataSnapshot.getChildrenCount());

                    Calendar c = Calendar.getInstance();
                    c.get(Calendar.DAY_OF_MONTH);
                    c.get(Calendar.MONTH);
                    c.get(Calendar.YEAR);
                    String fecha = DateFormat.getDateInstance(DateFormat.FULL).format(c.getTime());

                    HistoricoSemanalSubred hs = new HistoricoSemanalSubred();
                    hs.setId_historico(String.valueOf(maxidSubred+1));
                    hs.setTotal_asistencia(total_asistencia);
                    hs.setTotal_ofrenda(total_ofrenda);
                    hs.setFecha(fecha);

                    databaseReference.child("Historico Subred").child(hs.getId_historico()).setValue(hs);

                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });


    }


    private void crearFormato(final String [] fecha) {
        databaseReference.child("Registro Celula").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                int total_asistencia = 0;
                int total_invitados = 0;
                int total_ninos = 0;
                double total_ofrenda = 0;

                for (final DataSnapshot snapshot: dataSnapshot.getChildren()){
                    RegistroCelula rc = snapshot.getValue(RegistroCelula.class);
                    String fecha_celula = rc.getFecha_celula();

                    for (int i = 0; i < fecha.length; i++) {
                        if (fecha[i].equals(fecha_celula)) {
                            total_asistencia = total_asistencia + rc.getAsistencia_celula();
                            total_invitados = total_invitados + rc.getInvitados_celula();
                            total_ninos = total_ninos + rc.getNinos_celula();
                            total_ofrenda = total_ofrenda + rc.getOfrenda_celula();
                        }
                    }
                }

                totalSemanal(total_asistencia,total_invitados,total_ninos,total_ofrenda);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }
    private void totalSemanal(final int total_asistencia, final int total_invitados, final int total_ninos, final double total_ofrenda) {

        databaseReference.child("Historico").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if(dataSnapshot.exists()){
                    maxid=(dataSnapshot.getChildrenCount());
                    Toast.makeText(PrincipalActivity.this, String.valueOf(maxid), Toast.LENGTH_SHORT).show();

                    Calendar c = Calendar.getInstance();
                    c.get(Calendar.DAY_OF_MONTH);
                    c.get(Calendar.MONTH);
                    c.get(Calendar.YEAR);
                    String fecha = DateFormat.getDateInstance(DateFormat.FULL).format(c.getTime());

                    HistoricoSemanal hs3 = new HistoricoSemanal();
                    hs3.setId_historico(String.valueOf(maxid+1));
                    hs3.setTotal_asistencia(total_asistencia);
                    hs3.setTotal_invitados(total_invitados);
                    hs3.setTotal_ninos(total_ninos);
                    hs3.setTotal_ofrenda(total_ofrenda);
                    hs3.setFecha(fecha);

                    databaseReference.child("Historico").child(hs3.getId_historico()).setValue(hs3);
                }

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });


    }

    @Override
    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
        Bundle bundle = new Bundle();

        Calendar c = Calendar.getInstance();
        c.set(Calendar.YEAR, year);
        c.set(Calendar.MONTH, month);
        c.set(Calendar.DAY_OF_MONTH, dayOfMonth);
        String currentDateString = DateFormat.getDateInstance(DateFormat.FULL).format(c.getTime());


        c.setFirstDayOfWeek(Calendar.TUESDAY);

        Calendar c2 = Calendar.getInstance();
        c2.setFirstDayOfWeek(Calendar.TUESDAY);

        int numero_semana_original = c2.get(Calendar.WEEK_OF_YEAR);
        int numero_semana = c.get(Calendar.WEEK_OF_YEAR);


        System.out.println("El numero de la semana es: " + numero_semana);
        System.out.println("El numero de la semana de hoy es: " + numero_semana_original);

        if (numero_semana_original == numero_semana) {
            System.out.println(currentDateString);
            if (resvar == 1) {
                RegistroSubredFragment.setFecha(currentDateString);
            }
            if (resvar == 2) {
                RegistroCelulaFragment.setFecha(currentDateString);
            }

        } else {
            Toast.makeText(PrincipalActivity.this, "La fecha de la semana no coincide con el de hoy", Toast.LENGTH_LONG).show();
        }

        if (resvar == 3) {
            HistoricoDiarioFragment.setFecha(currentDateString);
        }
    }
}

File: app/src/main/java/com/david/redcristianauno/PrincipalActivity.kt
package com.david.redcristianauno

import android.app.AlertDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.DatePicker
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar

import com.david.redcristianauno.Firestore.InsertarDatos
import com.david.redcristianauno.Firestore.LeerDatos
import com.david.redcristianauno.Historico.HistoricoDiarioFragment
import com.david.redcristianauno.POJOs.HistoricoSemanal
import com.david.redcristianauno.POJOs.HistoricoSemanalSubred
import com.david.redcristianauno.POJOs.RegistroCelula
import com.david.redcristianauno.POJOs.RegistroSubred

import com.david.redcristianauno.ui.activities.LoginActivity
import com.david.redcristianauno.ui.fragments.BiblieFragment
import com.david.redcristianauno.ui.fragments.CreateConfigurationDialogFragment
import com.google.android.material.navigation.NavigationView
import com.google.firebase.FirebaseApp
import com.google.firebase.database.*
import java.text.DateFormat
import java.util.*

class PrincipalActivity : AppCompatActivity(),
    NavigationView.OnNavigationItemSelectedListener, OnDateSetListener {
    private var tv_nombre_usuario: TextView? = null
    private var tv_correo_usuario: TextView? = null
    private var firebaseDatabase: FirebaseDatabase? = null
    private var databaseReference: DatabaseReference? = null
    private val listaSemanal: List<String> =
        ArrayList()
    private var navigationView: NavigationView? = null
    private val menu: Menu? = null
    var resvar = 0
    var maxid: Long = 0
    var maxidSubred: Long = 0
    private val l = LeerDatos()
    private val inda = InsertarDatos()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //setContentView(R.layout.activity_principal)
        val toolbar =
            findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)



        val nombreview = navigationView?.getHeaderView(0)
        val correoview = navigationView?.getHeaderView(0)
        tv_nombre_usuario =
            nombreview?.findViewById<View>(R.id.tv_nombre_usuario) as TextView
        tv_correo_usuario = correoview?.findViewById<View>(R.id.tv_correo) as TextView
        navigationView?.setNavigationItemSelectedListener(this)
        val fragmentManager = supportFragmentManager
        fragmentManager.beginTransaction().replace(R.id.contenedor, MenuFragment()).commit()
        inicializarFirebase()
        val c = Calendar.getInstance()
        c[Calendar.DAY_OF_MONTH]
        c[Calendar.MONTH]
        c[Calendar.YEAR]
        val fecha =
            DateFormat.getDateInstance(DateFormat.FULL).format(c.time)

        //checar(fecha);
        inda.leerHistorico(fecha)
        correo_usuario =
            Preferences.obtenerPreferencesString(
                this,
                Preferences.PREFERENCES_USUARIO_LOGIN
            )
        tv_correo_usuario!!.text = correo_usuario
        l.datosSesion(
            tv_nombre_usuario,
            correo_usuario,
            menu,
            navigationView,
            this@PrincipalActivity
        )

        //ocultardatos();
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.principal, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.salida) {
            AlertDialog.Builder(this@PrincipalActivity)
                .setTitle("Cerrar Sesión")
                .setMessage("¿Estás seguro de cerrar la sesión?")
                .setPositiveButton("Aceptar") { dialog, which ->
                    Preferences.savePreferenceBoolean(
                        this@PrincipalActivity,
                        false,
                        Preferences.PREFENCE_ESTADO_BUTTON_SESION
                    )
                    val i = Intent(this@PrincipalActivity, LoginActivity::class.java)
                    startActivity(i)
                    finish()
                }
                .setNegativeButton("Cancelar") { dialog, which -> dialog.cancel() }.show()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId
        val fragmentManager = supportFragmentManager
        if (id == R.id.nav_home) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, MenuFragment()).commit()
        } else if (id == R.id.nav_gallery) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, RegistroSubredFragment())
                .commit()
            resvar = 1
        } else if (id == R.id.nav_slideshow) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, RegistroCelulaFragment())
                .commit()
            resvar = 2
        } else if (id == R.id.nav_tools) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, HistoricoFragment())
                .commit()
            resvar = 3
        } else if (id == R.id.nav_biblia) {
            fragmentManager.beginTransaction().replace(R.id.contenedor,
                BiblieFragment()
            ).commit()
        } else if (id == R.id.nav_share) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, PermisosUsuariosFragment())
                .commit()
        } else if (id == R.id.crear) {
            fragmentManager.beginTransaction()
                .replace(R.id.contenedor, CreateConfigurationDialogFragment()).commit()
        } else if (id == R.id.noticias) {
            fragmentManager.beginTransaction().replace(R.id.contenedor, noticiasFragment()).commit()
        }

        return true
    }

    fun inicializarFirebase() {
        FirebaseApp.initializeApp(this)
        firebaseDatabase = FirebaseDatabase.getInstance()
        databaseReference = firebaseDatabase!!.reference
    }

    private fun checar(fecha: String) {
        databaseReference!!.child("Historico").orderByChild("fecha")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                var bandera = true
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    for (snapshot in dataSnapshot.children) {
                        val rc = snapshot.getValue(
                            HistoricoSemanal::class.java
                        )
                        val fecha_celula = rc!!.fecha
                        println("Fecha historico:$fecha_celula")
                        if (fecha == fecha_celula) {
                            bandera = false
                        }
                        println("Bandera: $bandera")
                    }
                    if (bandera == true) {
                        formatoSemanal()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    fun formatoSemanal() {
        val miCalendario = Calendar.getInstance()
        miCalendario.firstDayOfWeek = Calendar.TUESDAY
        var bandera = false
        val diaSemana = miCalendario[Calendar.DAY_OF_WEEK]
        bandera = if (diaSemana != miCalendario.firstDayOfWeek) {
            false
        } else {
            true
        }
        if (bandera == true) {
            val dia_del_mes = miCalendario[Calendar.DAY_OF_MONTH]
            println("El dia es: $dia_del_mes")
            var mes = miCalendario[Calendar.MONTH] + 1
            println("El mes es: $mes")
            val año = miCalendario[Calendar.YEAR]
            println("El año es: $año")
            var nuevo = dia_del_mes
            val fechas = arrayOfNulls<String>(7)
            for (i in 0..6) {
                if (nuevo == 0) {
                    nuevo = 1
                }
                nuevo -= 1
                when (mes) {
                    1 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    2 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    3 -> if (año % 4 == 0 && (año % 100 != 0 || año % 400 == 0)) {
                        if (nuevo <= 0) {
                            mes = mes - 1
                            nuevo = 29 - nuevo
                        }
                    } else {
                        if (nuevo <= 0) {
                            mes = mes - 1
                            nuevo = 28 - nuevo
                        }
                    }
                    4 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    5 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 30 - nuevo
                    }
                    6 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    7 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 30 - nuevo
                    }
                    8 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    9 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    10 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 30 - nuevo
                    }
                    11 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 31 - nuevo
                    }
                    12 -> if (nuevo <= 0) {
                        mes = mes - 1
                        nuevo = 30 - nuevo
                    }
                }
                miCalendario[Calendar.DAY_OF_MONTH] = nuevo
                miCalendario[Calendar.MONTH] = mes - 1
                miCalendario[Calendar.YEAR] = año
                val fecha =
                    DateFormat.getDateInstance(DateFormat.FULL)
                        .format(miCalendario.time)
                println(fecha)
                fechas[i] = fecha
            }
            crearFormato(fechas)
            crearFormatoSubred(fechas)
        }
    }

    private fun crearFormatoSubred(fechas: Array<String?>) {
        databaseReference!!.child("Registro Subred")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    var total_asistencia = 0
                    var total_ofrenda = 0.0
                    for (snapshot in dataSnapshot.children) {
                        val rc =
                            snapshot.getValue(RegistroSubred::class.java)
                        val fecha_celula = rc!!.fecha_subred
                        for (i in fechas.indices) {
                            if (fechas[i] == fecha_celula) {
                                total_asistencia = total_asistencia + rc.asistencia_subred
                                total_ofrenda = total_ofrenda + rc.ofrenda_subred
                            }
                        }
                    }
                    totalSemanalSubred(total_asistencia, total_ofrenda)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    private fun totalSemanalSubred(
        total_asistencia: Int,
        total_ofrenda: Double
    ) {
        databaseReference!!.child("Historico Subred")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists()) {
                        maxidSubred = dataSnapshot.childrenCount
                        val c = Calendar.getInstance()
                        c[Calendar.DAY_OF_MONTH]
                        c[Calendar.MONTH]
                        c[Calendar.YEAR]
                        val fecha =
                            DateFormat.getDateInstance(DateFormat.FULL)
                                .format(c.time)
                        val hs = HistoricoSemanalSubred()
                        hs.id_historico = (maxidSubred + 1).toString()
                        hs.total_asistencia = total_asistencia
                        hs.total_ofrenda = total_ofrenda
                        hs.fecha = fecha
                        databaseReference!!.child("Historico Subred").child(hs.id_historico)
                            .setValue(hs)
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    private fun crearFormato(fecha: Array<String?>) {
        databaseReference!!.child("Registro Celula")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    var total_asistencia = 0
                    var total_invitados = 0
                    var total_ninos = 0
                    var total_ofrenda = 0.0
                    for (snapshot in dataSnapshot.children) {
                        val rc =
                            snapshot.getValue(RegistroCelula::class.java)
                        val fecha_celula = rc!!.fecha_celula
                        for (i in fecha.indices) {
                            if (fecha[i] == fecha_celula) {
                                total_asistencia = total_asistencia + rc.asistencia_celula
                                total_invitados = total_invitados + rc.invitados_celula
                                total_ninos = total_ninos + rc.ninos_celula
                                total_ofrenda = total_ofrenda + rc.ofrenda_celula
                            }
                        }
                    }
                    totalSemanal(total_asistencia, total_invitados, total_ninos, total_ofrenda)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    private fun totalSemanal(
        total_asistencia: Int,
        total_invitados: Int,
        total_ninos: Int,
        total_ofrenda: Double
    ) {
        databaseReference!!.child("Historico")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists()) {
                        maxid = dataSnapshot.childrenCount
                        Toast.makeText(this@PrincipalActivity, maxid.toString(), Toast.LENGTH_SHORT)
                            .show()
                        val c = Calendar.getInstance()
                        c[Calendar.DAY_OF_MONTH]
                        c[Calendar.MONTH]
                        c[Calendar.YEAR]
                        val fecha =
                            DateFormat.getDateInstance(DateFormat.FULL)
                                .format(c.time)
                        val hs3 = HistoricoSemanal()
                        hs3.id_historico = (maxid + 1).toString()
                        hs3.total_asistencia = total_asistencia
                        hs3.total_invitados = total_invitados
                        hs3.total_ninos = total_ninos
                        hs3.total_ofrenda = total_ofrenda
                        hs3.fecha = fecha
                        databaseReference!!.child("Historico").child(hs3.id_historico)
                            .setValue(hs3)
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    override fun onDateSet(
        view: DatePicker,
        year: Int,
        month: Int,
        dayOfMonth: Int
    ) {
        val bundle = Bundle()
        val c = Calendar.getInstance()
        c[Calendar.YEAR] = year
        c[Calendar.MONTH] = month
        c[Calendar.DAY_OF_MONTH] = dayOfMonth
        val currentDateString =
            DateFormat.getDateInstance(DateFormat.FULL).format(c.time)
        c.firstDayOfWeek = Calendar.TUESDAY
        val c2 = Calendar.getInstance()
        c2.firstDayOfWeek = Calendar.TUESDAY
        val numero_semana_original = c2[Calendar.WEEK_OF_YEAR]
        val numero_semana = c[Calendar.WEEK_OF_YEAR]
        println("El numero de la semana es: $numero_semana")
        println("El numero de la semana de hoy es: $numero_semana_original")
        if (numero_semana_original == numero_semana) {
            println(currentDateString)
            if (resvar == 1) {
                RegistroSubredFragment.setFecha(currentDateString)
            }
            if (resvar == 2) {
                RegistroCelulaFragment.setFecha(currentDateString)
            }
        } else {
            Toast.makeText(
                this@PrincipalActivity,
                "La fecha de la semana no coincide con el de hoy",
                Toast.LENGTH_LONG
            ).show()
        }
        if (resvar == 3) {
            HistoricoDiarioFragment.setFecha(currentDateString)
        }
    }

    companion object {
        var asistencia = 0
        var invitados = 0
        var ninos = 0
        var ofrenda = 0.0
        var correo_usuario = ""
        var res = ""
    }
}